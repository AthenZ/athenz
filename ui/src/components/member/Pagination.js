/*
 * Copyright The Athenz Authors
 * Licensed under the Apache License, Version 2.0
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import Button from '../denali/Button';
import { colors } from '../denali/styles';

const PaginationContainer = styled.div`
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin: 15px 0;
    padding: 10px 0;
    border-top: 1px solid ${colors.grey300};
`;

const PageInfo = styled.span`
    color: ${colors.grey700};
    font-size: 14px;
    font-weight: 500;
`;

const PageButtons = styled.div`
    display: flex;
    gap: 5px;
    align-items: center;
`;

const PageButton = styled(Button)`
    min-width: 36px;
    height: 32px;
    padding: 0 8px;
`;

export default class Pagination extends React.Component {
    generatePageNumbers() {
        const { currentPage, totalPages } = this.props;
        const maxVisible = 9;
        
        if (totalPages <= maxVisible) {
            return Array.from({ length: totalPages }, (_, i) => i + 1);
        }
        
        // Dynamic page number logic for max 9 pages
        const start = Math.max(1, Math.min(currentPage - 4, totalPages - maxVisible + 1));
        const end = Math.min(totalPages, start + maxVisible - 1);
        
        return Array.from({ length: end - start + 1 }, (_, i) => start + i);
    }
    
    render() {
        const { currentPage, totalPages, onPageChange } = this.props;
        
        if (totalPages <= 1) {
            return null;
        }
        
        const pageNumbers = this.generatePageNumbers();
        
        return (
            <PaginationContainer data-testid="pagination">
                <PageInfo>
                    Page {currentPage} / {totalPages}
                </PageInfo>
                
                <PageButtons>
                    <PageButton 
                        disabled={currentPage === 1}
                        onClick={() => onPageChange(1)}
                        size="small"
                        variant="secondary"
                    >
                        First
                    </PageButton>
                    
                    <PageButton
                        disabled={currentPage === 1}
                        onClick={() => onPageChange(currentPage - 1)}
                        size="small"
                        variant="secondary"
                    >
                        ‹
                    </PageButton>
                    
                    {pageNumbers.map(pageNumber => (
                        <PageButton
                            key={pageNumber}
                            variant={currentPage === pageNumber ? 'primary' : 'secondary'}
                            onClick={() => onPageChange(pageNumber)}
                            size="small"
                        >
                            {pageNumber}
                        </PageButton>
                    ))}
                    
                    <PageButton
                        disabled={currentPage === totalPages}
                        onClick={() => onPageChange(currentPage + 1)}
                        size="small"
                        variant="secondary"
                    >
                        ›
                    </PageButton>
                    
                    <PageButton 
                        disabled={currentPage === totalPages}
                        onClick={() => onPageChange(totalPages)}
                        size="small"
                        variant="secondary"
                    >
                        Last
                    </PageButton>
                </PageButtons>
            </PaginationContainer>
        );
    }
}

Pagination.propTypes = {
    /** Current page number */
    currentPage: PropTypes.number.isRequired,
    /** Total number of pages */
    totalPages: PropTypes.number.isRequired,
    /** Page change handler */
    onPageChange: PropTypes.func.isRequired,
};