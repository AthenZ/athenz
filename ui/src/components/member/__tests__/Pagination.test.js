/*
 * Copyright The Athenz Authors
 * Licensed under the Apache License, Version 2.0
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Pagination from '../Pagination';

describe('Pagination', () => {
    const defaultProps = {
        currentPage: 1,
        totalPages: 5,
        onPageChange: jest.fn(),
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should render without errors', () => {
        const { getByTestId } = render(<Pagination {...defaultProps} />);
        expect(getByTestId('pagination')).toBeInTheDocument();
    });

    it('should display current page and total pages', () => {
        const { getByText } = render(<Pagination {...defaultProps} />);
        expect(getByText('Page 1 / 5')).toBeInTheDocument();
    });

    it('should not render when totalPages is 1 or less', () => {
        const { container } = render(
            <Pagination {...defaultProps} totalPages={1} />
        );
        expect(container.firstChild).toBeNull();
    });

    it('should render all page numbers when totalPages <= 9', () => {
        const { getByText } = render(
            <Pagination {...defaultProps} totalPages={7} />
        );
        
        for (let i = 1; i <= 7; i++) {
            expect(getByText(i.toString())).toBeInTheDocument();
        }
    });

    it('should render maximum 9 page numbers when totalPages > 9', () => {
        const { container } = render(
            <Pagination {...defaultProps} totalPages={15} />
        );
        
        // Should show pages 1-9 when current page is 1
        const pageButtons = container.querySelectorAll('button');
        const numberButtons = Array.from(pageButtons).filter(button => 
            /^\d+$/.test(button.textContent.trim())
        );
        expect(numberButtons).toHaveLength(9);
    });

    it('should handle first page navigation', () => {
        const onPageChange = jest.fn();
        const { getByText } = render(
            <Pagination 
                currentPage={3} 
                totalPages={5} 
                onPageChange={onPageChange} 
            />
        );
        
        fireEvent.click(getByText('First'));
        expect(onPageChange).toHaveBeenCalledWith(1);
    });

    it('should handle last page navigation', () => {
        const onPageChange = jest.fn();
        const { getByText } = render(
            <Pagination 
                currentPage={3} 
                totalPages={5} 
                onPageChange={onPageChange} 
            />
        );
        
        fireEvent.click(getByText('Last'));
        expect(onPageChange).toHaveBeenCalledWith(5);
    });

    it('should handle previous page navigation', () => {
        const onPageChange = jest.fn();
        const { getByText } = render(
            <Pagination 
                currentPage={3} 
                totalPages={5} 
                onPageChange={onPageChange} 
            />
        );
        
        fireEvent.click(getByText('‹'));
        expect(onPageChange).toHaveBeenCalledWith(2);
    });

    it('should handle next page navigation', () => {
        const onPageChange = jest.fn();
        const { getByText } = render(
            <Pagination 
                currentPage={3} 
                totalPages={5} 
                onPageChange={onPageChange} 
            />
        );
        
        fireEvent.click(getByText('›'));
        expect(onPageChange).toHaveBeenCalledWith(4);
    });

    it('should handle direct page number click', () => {
        const onPageChange = jest.fn();
        const { getByText } = render(
            <Pagination 
                currentPage={1} 
                totalPages={5} 
                onPageChange={onPageChange} 
            />
        );
        
        fireEvent.click(getByText('3'));
        expect(onPageChange).toHaveBeenCalledWith(3);
    });

    it('should disable first/prev buttons on first page', () => {
        const { getByText } = render(
            <Pagination 
                currentPage={1} 
                totalPages={5} 
                onPageChange={jest.fn()} 
            />
        );
        
        expect(getByText('First')).toBeDisabled();
        expect(getByText('‹')).toBeDisabled();
    });

    it('should disable last/next buttons on last page', () => {
        const { getByText } = render(
            <Pagination 
                currentPage={5} 
                totalPages={5} 
                onPageChange={jest.fn()} 
            />
        );
        
        expect(getByText('Last')).toBeDisabled();
        expect(getByText('›')).toBeDisabled();
    });

    it('should highlight current page button', () => {
        const { getByText } = render(
            <Pagination 
                currentPage={3} 
                totalPages={5} 
                onPageChange={jest.fn()} 
            />
        );
        
        const currentPageButton = getByText('3');
        // Button should have primary variant (this depends on Button implementation)
        expect(currentPageButton).toBeInTheDocument();
    });

    it('should generate correct page numbers for middle pages', () => {
        // When current page is in the middle of a large set
        const { getByText } = render(
            <Pagination 
                currentPage={10} 
                totalPages={20} 
                onPageChange={jest.fn()} 
            />
        );
        
        // Should show pages around current page (6-14 when current is 10)
        expect(getByText('6')).toBeInTheDocument();
        expect(getByText('10')).toBeInTheDocument();
        expect(getByText('14')).toBeInTheDocument();
    });

    it('should generate correct page numbers for end pages', () => {
        // When current page is near the end
        const { getByText } = render(
            <Pagination 
                currentPage={18} 
                totalPages={20} 
                onPageChange={jest.fn()} 
            />
        );
        
        // Should show last 9 pages (12-20)
        expect(getByText('12')).toBeInTheDocument();
        expect(getByText('18')).toBeInTheDocument();
        expect(getByText('20')).toBeInTheDocument();
    });

    it('should match snapshot', () => {
        const { container } = render(<Pagination {...defaultProps} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot with many pages', () => {
        const { container } = render(
            <Pagination 
                currentPage={10} 
                totalPages={20} 
                onPageChange={jest.fn()} 
            />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});