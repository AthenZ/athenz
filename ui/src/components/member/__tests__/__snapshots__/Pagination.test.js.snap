// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should match snapshot 1`] = `
.emotion-0 {
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  margin: 15px 0;
  padding: 10px 0;
  border-top: 1px solid #f2f2f2;
}

.emotion-2 {
  color: #606060;
  font-size: 14px;
  font-weight: 500;
}

.emotion-4 {
  display: flex;
  gap: 5px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-6 {
  border: none;
  border-radius: 2px;
  box-shadow: 0 0 0 1px transparent inset,0 0 0 0 #d5d5d5 inset;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  font-family: Helvetica,Arial,sans-serif;
  font-weight: 300;
  font-size: 12px;
  line-height: 1;
  margin: 5px;
  min-width: 70px;
  outline: 0;
  padding: 8px 18px;
  text-align: center;
  text-shadow: none;
  text-transform: none;
  -webkit-transition: all 0.2s ease-in;
  transition: all 0.2s ease-in;
  vertical-align: baseline;
  white-space: nowrap;
  background: linear-gradient(
              to right,
              #3697f2,
              #3570f4
          );
  color: #fff;
}

.emotion-6:first-of-type {
  margin-left: 0;
}

.emotion-6:disabled {
  background: rgba(48,48,48,0.1);
  color: rgba(48,48,48,0.2);
  cursor: not-allowed;
}

.emotion-6:hover:not(:disabled) {
  background: linear-gradient(
                  to right,
                  #3570f4,
                  #3448f7
              );
}

.emotion-7 {
  min-width: 36px;
  height: 32px;
  padding: 0 8px;
}

<div
  class="emotion-0 emotion-1"
  data-testid="pagination"
>
  <span
    class="emotion-2 emotion-3"
  >
    Page 
    1
     / 
    5
  </span>
  <div
    class="emotion-4 emotion-5"
  >
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      disabled=""
      variant="secondary"
    >
      First
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      disabled=""
      variant="secondary"
    >
      ‹
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="primary"
    >
      1
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      2
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      3
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      4
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      5
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      ›
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      Last
    </button>
  </div>
</div>
`;

exports[`Pagination should match snapshot with many pages 1`] = `
.emotion-0 {
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  margin: 15px 0;
  padding: 10px 0;
  border-top: 1px solid #f2f2f2;
}

.emotion-2 {
  color: #606060;
  font-size: 14px;
  font-weight: 500;
}

.emotion-4 {
  display: flex;
  gap: 5px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-6 {
  border: none;
  border-radius: 2px;
  box-shadow: 0 0 0 1px transparent inset,0 0 0 0 #d5d5d5 inset;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  font-family: Helvetica,Arial,sans-serif;
  font-weight: 300;
  font-size: 12px;
  line-height: 1;
  margin: 5px;
  min-width: 70px;
  outline: 0;
  padding: 8px 18px;
  text-align: center;
  text-shadow: none;
  text-transform: none;
  -webkit-transition: all 0.2s ease-in;
  transition: all 0.2s ease-in;
  vertical-align: baseline;
  white-space: nowrap;
  background: linear-gradient(
              to right,
              #3697f2,
              #3570f4
          );
  color: #fff;
}

.emotion-6:first-of-type {
  margin-left: 0;
}

.emotion-6:disabled {
  background: rgba(48,48,48,0.1);
  color: rgba(48,48,48,0.2);
  cursor: not-allowed;
}

.emotion-6:hover:not(:disabled) {
  background: linear-gradient(
                  to right,
                  #3570f4,
                  #3448f7
              );
}

.emotion-7 {
  min-width: 36px;
  height: 32px;
  padding: 0 8px;
}

<div
  class="emotion-0 emotion-1"
  data-testid="pagination"
>
  <span
    class="emotion-2 emotion-3"
  >
    Page 
    10
     / 
    20
  </span>
  <div
    class="emotion-4 emotion-5"
  >
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      First
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      ‹
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      6
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      7
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      8
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      9
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="primary"
    >
      10
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      11
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      12
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      13
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      14
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      ›
    </button>
    <button
      class="emotion-6 denali-button emotion-7 emotion-8"
      variant="secondary"
    >
      Last
    </button>
  </div>
</div>
`;
